benchmarks/encryption.py

```diff
@@ -0,0 +1,330 @@
+                hmac_salt = enc[32:48]
+
+                key = self._derive_key(salt)
+                hmac_key = self._derive_key(hmac_salt)
+
+                h = self._hmac(
+                    hmac_key,
+                    self._hash(),
+                    backend=self._backend()
+                )
+
+                ciph = self._cipher(
+                    self._algorithm(key),
+                    self._mode(iv),
+                    backend=self._backend()
+                )
+
+                decryptor = ciph.decryptor()
+                h.update(iv + salt + hmac_salt)
+
+                enc = enc[48:]
+
+            if nextenc is None:
+                signature = enc[-32:]
+                enc = enc[:-32]
+                h.update(enc)
+
+                h2 = h.copy()
+                h.verify(signature)
+                h2.finalize()
+
+            else:
+                h.update(enc)
+
+            yield decryptor.update(enc)
+
+        yield decryptor.finalize()
+
+
+class _Encryption(ABC):
+
+    @property
+    def extension(self)->str:
+        raise NotImplementedError
+
+    @contextmanager
+    def encrypt(self, fobj: IO[bytes]) -> IO[bytes]:
+        raise NotImplementedError
+
+    @contextmanager
+    def deserialize(self, fobj: IO[bytes]) -> IO[bytes]:
+        raise NotImplementedError
+
+
+class AesEncryption(_Encryption):
+
+    extension = 'aes'
+
+    @contextmanager
+    def encrypt(self, fobj: IO[bytes]) -> IO[bytes]:
+        salt = urandom(16)
+        hmac_salt = urandom(16)
+
+        key_deriver = KeyDerive(
+            b"client10",
+            PASSWORD_DERIVE_ITER,
+            SHA256,
+            PBKDF2HMAC,
+            default_backend,
+            salt
+        )
+
+        hmac_key_deriver = KeyDerive(
+            b"client10",
+            PASSWORD_DERIVE_ITER,
+            SHA256,
+            PBKDF2HMAC,
+            default_backend,
+            hmac_salt
+        )
+
+        key = key_deriver()
+        hmac_key = hmac_key_deriver()
+
+        encrypt_stream = Encrypt(
+        key,
+        hmac_key,
+        HMAC,
+        SHA256,
+        Cipher,
+        AES,
+        CTR,
+        default_backend,
+        salt,
+        hmac_salt
+    )
+
+        try:
+            with NamedTemporaryFile() as temp:
+                yield temp
+                temp.seek(0)
+                for item in encrypt_stream(temp):
+                    fobj.write(item)
+
+        finally:
+            pass
+
+    @contextmanager
+    def deserialize(self, fobj: IO[bytes]) -> IO[bytes]:
+
+        decrypt = Decrypt(
+            b"client10",
+            SHA256,
+            PBKDF2HMAC,
+            HMAC,
+            Cipher,
+            AES,
+            CTR,
+            default_backend
+        )
+
+        try:
+            with NamedTemporaryFile() as temp:
+                for item in decrypt(fobj):
+                    temp.write(item)
+                yield temp
+
+        finally:
```

Comment:
Is this finally block (and the one above) required since it doesn't execute any code?

Reply:
Not needed at all. Will remove!